src/main/python/SystemML/defmatrix.py

```diff
@@ -0,0 +1,285 @@
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+import numpy as np
+
+from . import pydml, MLContext
+from .converters import *
+from pyspark import SparkContext, RDD
+from pyspark.sql import DataFrame, SQLContext
+
+def setSparkContext(sc):
+    """
+    Before using the matrix, the user needs to invoke this function.
+    
+    Parameters
+    ----------
+    sc: SparkContext
+        SparkContext
+    """
+    matrix.ml = MLContext(sc)
+    matrix.sc = sc
+    
+def checkIfMLContextIsSet():
+    if matrix.ml is None:
+        raise Exception('Expected setSparkContext(sc) to be called.')
+
+class DMLOp(object):
+    def __init__(self, inputs, dml=None):
+        self.inputs = inputs
+        self.dml = dml
+        
+    def _visit(self, execute=True):
+        matrix.dml = matrix.dml + self.dml
+            
+
+def reset():
+    for m in matrix.visited:
+        m.visited = False
+    matrix.visited = []
+        
+def binaryOp(lhs, rhs, opStr):
+    inputs = []
+    if isinstance(lhs, matrix):
+        lhsStr = lhs.ID
+        inputs = [lhs]
+    elif isinstance(lhs, float) or isinstance(lhs, int):
+        lhsStr = str(lhs)
+    else:
+        raise TypeError('Incorrect type')
+    if isinstance(rhs, matrix):
+        rhsStr = rhs.ID
+        inputs = inputs + [rhs]
+    elif isinstance(rhs, float) or isinstance(rhs, int):
+        rhsStr = str(rhs)
+    else:
+        raise TypeError('Incorrect type')
+    dmlOp = DMLOp(inputs)
+    out = matrix(None, op=dmlOp)
+    dmlOp.dml = [out.ID, ' = ', lhsStr, opStr, rhsStr, '\n']
+    return out
+
+def eval(outputs, outputDF=False, execute=True):
+    """
+    Executes the unevaluated DML script and computes the matrices specified by outputs.
+
+    Parameters
+    ----------
+    outputs: list of matrices
+    outputDF: back the data of matrix as PySpark DataFrame
+    """
+    checkIfMLContextIsSet()
+    reset()
+    matrix.dml = []
+    matrix.script = pydml('')
+    if isinstance(outputs, matrix):
+        outputs = [ outputs ]
+    elif not isinstance(outputs, list):
+        raise TypeError('Incorrect input type')
+    for m in outputs:
+        m.output = True
+        m._visit(execute=execute)
+    if not execute:
+        return ''.join(matrix.dml)
+    matrix.script.scriptString = ''.join(matrix.dml)
+    results = matrix.ml.execute(matrix.script)
+    for m in outputs:
+        if outputDF:
+            m.data = results.getDataFrame(m.ID)
+        else:
+            m.data = results.getNumPyArray(m.ID)
```

Comment:
First, this lazy evaluation of the new `matrix` class is awesome!

Continuing with my thought on fusing the Python MLContext `Matrix` and `matrix` classes.  The current `Matrix` class has a few cool properties -- it can be used directly in other scripts as inputs (i.e. it represents a fully evaluated matrix), and at the same time it can be converted to other types as desired by the user, i.e. to a `DataFrame` via `toDF()`, and to NumPy arrays once I add the function.  If we merge that `Matrix` class with the new `matrix` class (using the later name), these lazy matrices could be used in either future direct matrix manipulation via this new DSL, *as well as* future scripts directly via `MLContext`.  Once a user wanted the results, he/she simply could call `m1.toDF()`, `m1.toNumPyArray()`, etc.

This lazy matrix DSL is exciting!